### ISSUE TITLE
Add support for more intrinsics

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/fparser#428]
One socrates file fails on an unknown symbol (`erf`). PSyclone has support for the `ERF` `InstrincCall` now, but its missing in fparser. Is it possible to add it to https://github.com/stfc/fparser/blob/5018a64e765187b444a3e3540eb2c0dd10262f24/src/fparser/two/Fortran2003.py#L12027 ?

**Comment by hiker** [comment]
There is also #434 - related to some(?) intrinsics not properly recognised. Might be worth trying to fix this at the same time?

**Comment by tinyendian** [comment]
Thanks @hiker - I did come across a case where PSyclone complained that Fortran 2008 intrinsic [gamma](https://gcc.gnu.org/onlinedocs/gfortran/GAMMA.html) is undeclared:
```bash
psyclone.psyir.backend.visitor.VisitorError: Visitor Error: The following symbols are not explicitly declared or imported from a module and there are no wildcard imports which could be bringing them into scope: 'gamma'
```
Tentatively adding `gamma` to dict `mathematical_names` in `fparser/two/Fortran2003.py` resolved the issue, but it might not be the right way to do it as @hiker said, given that `gamma` was introduced by Fortran 2008.

**Comment by LonelyCat124** [comment]
We also hit ERF not being recognized for one UKCA file @RobWatersMet . I will have a look at how F2008 is supported in fparser and try to add something for the F2008 intrinsics listed here.

**Comment by LonelyCat124** [comment]
So I think what I could do is in Fortran2008.py I could add:
```
class Fortran2008_Intrinsic_Name(Intrinsic_Name):
"""
...
"""

f08_math_intrinsics = {
    "ERF": {"min": 1, "max": 1},
    "GAMMA": {"min": 1, "max": 1},
}

f08_bitshift_intrinsics = {
         SHIFTL": {"min": 2, "max": 2},
        "SHIFTR": {"min": 2, "max": 2},
        "SHIFTA": {"min": 2, "max": 2},
}

    generic_function_names = {}
    generic_function_names.update(f08_math_intrinsics)
    generic_function_names.update(f08_bitshift_intrinsics)

    # A list of all function names
    function_names = list(generic_function_names.keys()) + list(
        specific_function_names.keys()
    )

def match():
...
```

I'm not sure what the subclass_names method on Fparser node types means - if anyone knows maybe there's a way to do it using that,. I'm not sure I actually need the base intrinsic names here.

I will not be adding the nonstandard intrinsics yet as I don't have a solution yet.

**Comment by arporter** [comment]
Happily the developer guide does actually describe subclass_names: https://fparser.readthedocs.io/en/latest/developers_guide.html#classes
I'm not sure whether that's what you need here though - you probably just need to copy/extend whatever Fortran2003.py does.

**Comment by arporter** [comment]
We do support some language extensions (i.e. non-standard features) but I can't remember offhand whether that includes any intrinsics.

**Comment by LonelyCat124** [comment]
Yeah, my issue is now in `Intrinsic_Function_Reference` unfortunately - it uses (hardcoded) `Intrinsic_Name.generic_function_names`, which then crashes on my example with ERF becuase ERF is defined in the F2008 subclass. This is solvable in theory, but I don't think I can check which language standard is being supported from inside `match` which is the headache. It might not matter though as I think we'd only get here if we'd matched to the F2008 anyway so it might be fine.

**Comment by LonelyCat124** [comment]
How this all works is a mystery, the `subclass_names` is doing some real wizardry and I don't quite understand it right now.


If I have `Fortran2008_Intrinsic_Name` in my subclass_names for `Intrinsic_Name`, then somehow (magically) we match "erf" as a lower case `Name` class in `Intrinsic_Function_Reference.match`, despite the `Fortran2008_Intrinsic_Name.match` never being called (which I know because it has a failing import in at the moment if you ever reach it), and also it should be a `Fortran2008_Intrinsic_Name` class (matching the usually behaviour, which is to get an `Intrinsic_Name` object in result).

I think potentially what I might need to do is change from staticmethod to classmethod on Intrinsic_Name.match, but I'm not yet sure. If I do that then I can do `type(result[0]).specific_function_names` and friends instead of using it explicitly, but its confusing me right now,. I'll probably look into this more tomorrow when I'm back at home and have more time.

**Comment by LonelyCat124** [comment]
Ok so I tracked down the issue right now and it is super confusing. I created fparser.two.Fortran2008.Fortran2008_Intrinsic_Name, but whatever is happening in the __init__.py in that directory creates a fparser.two.Fortran2003.Fortran2008_Intrinsic_Name and this doesn't keep the implementation of `match` from my implementation, but instead magics a new class which uses the `Name` class instead aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaah. I can probably fix it by renaming my class to end with something else.

**Comment by arporter** [comment]
@LonelyCat124 If/when you learn valuable things, could you also add them to the Dev Guide (if not already there) so we capture them?

**Comment by LonelyCat124** [comment]
Yeah. 

I thought a bit more on the bus and I really am not super keen on this solution (an Intrinsic_Name subclass), but I think due to all the lists/dicts in Intrinsic_Name being import time the nicer solution seems bad. It would be better to have some f2008 intrinsic dicts in the fortran2008 folder and append them if F2008 (or later) was specified as the standard, but I am not sure of a Pythonic way to do this right now, so I will go with this solution when I get it working tomorrow.

If the `subclass_names` magically solves my implementation then I will add it as an example of subclass names in the dev guide - I hope it does works as otherwise the implementation may be hard work (as I don't know the Fortran standard at the time I need it...).

The weird stuff thats going on in `fparser.two.Fortran2008.__init__.py` I am going to ignore and hope it goes away as it seems to redefine some of the F2008 classes as implementations of their base and the relevant match function, but then it loses anything defined on the F2008 method, but I also am not sure because the default (and what I'd been playing with) is F2003 (I'd assumed F2008) so I need to investigate more, maybe it doesn't overwrite the things with F2003, though it still means I get weirdness that anything ending _Name will generate objects in F2003 and give potentially wrong or bad errors (as it does for the overridden Intrinsic_Name).

**Comment by arporter** [comment]
Closing this as all F2008 intrinsics are now supported thanks to @LonelyCat124.
